### PR TITLE
Fixed #35038 - Changing violation_error_message on constraints cause a remove/add operation in migration

### PR DESCRIPTION
**Pull request by adriennefranke**
Fixes [#35038](https://code.djangoproject.com/ticket/35038). I created a method within the `BaseConstraint` class that returns a copy of the Constraint object without the two fields (`violation_error_code` and `violation_error_message`) that do not affect the DDL. I then call this method in `create_altered_constraints` and use that to determine when a migration for constraints is truly necessary.

There was another PR that was opened to fix this ticket so [here's a link](https://github.com/django/django/pull/17644) to that for reference.

### CODE REVIEW

**Comment by charettes** [comment]
Just like `verbose_name` changes on `Field` result in `AlterField` we'll want to track and autodetect these changes as well.

It is not already a noop at the schema application level we should make one but all attributes should be tracked.

**Comment by adriennefranke** [comment]
@charettes Thanks for explaining. I don't want to introduce code that's different in its design from the rest of the migration code so I'm happy to work on `AlterConstraint` if you think that's the best path forward.

**Comment by charettes** [comment]
@adriennefranke I think it warrants further discussions that I didn't see mentioned on the ticket so I figured I'd chime in.

Whether or not adding `AlterConstraint` is worth the trouble to keep things coherent around the code base should at least be discussed with a larger audience as we've taken an explicit stance with regards to options that don't affect underyling DDL in the past and I don't think that this aspect was considered when `Constraint`, `AddConstraint`, `RemoveConstraint` was considered.

**Comment by adriennefranke** [comment]
@charettes What would you say the right next steps are? Should I bring it up on the Discord or in the forums?

**Comment by charettes** [comment]
I think the forum would be a good location to start a discussion on that yes.

File: django/db/models/constraints.py [hunk 1]
```diff
@@ -89,6 +90,12 @@ def clone(self):
         _, args, kwargs = self.deconstruct()
         return self.__class__(*args, **kwargs)
 
+    def only_ddl_impacted_fields(self):
+        mutable_self = copy.copy(self)
+        mutable_self.violation_error_code = None
+        mutable_self.violation_error_message = None
+        return mutable_self
```
Comment (50-Course) [inline django/db/models/constraints.py:97]: Hi @adriennefranke, first timer on django core - hopped here from the djangocon US talks.

Thanks for the patch, and think its really an interesting one following the trend of conversations from #17644 and #13488.

In addition, to @charettes, could I suggest sending in a discourse update on the regarding this patch to discord as well?

Comment (adriennefranke) [reply django/db/models/constraints.py:97]: @50-Course Yes, I will do that as well. Thanks!

**Comment by felixxm** [comment]
> It seems like we should have an `AlterConstraint` operation that is in charge of dealing with that but it doesn't exist already 😬

Yes, yes, yes, we need `AlterConstraint` clearly documented as a no-op operation.

**Comment by adriennefranke** [comment]
Going to close this PR while I work on the `AlterConstraint` path. Thanks for the guidance everyone.

**Pull request closed without merging**